pageserver/src/walingest.rs

```diff
@@ -400,39 +398,72 @@ impl WalIngest {
-                            dbnode: decoded.blocks[1].rnode_dbnode,
-                            relnode: decoded.blocks[1].rnode_relnode,
-                        }),
-                        decoded.blocks[1].blkno / pg_constants::HEAPBLOCKS_PER_PAGE as u32,
-                        ZenithWalRecord::ClearVisibilityMapFlags {
-                            heap_blkno: decoded.blocks[1].blkno,
-                            flags: pg_constants::VISIBILITYMAP_VALID_BITS,
-                        },
-                    )?;
+                    if (xlrec.flags
+                        & (pg_constants::XLH_UPDATE_NEW_ALL_VISIBLE_CLEARED
+                            | pg_constants::XLH_UPDATE_OLD_ALL_VISIBLE_CLEARED))
+                        == (pg_constants::XLH_UPDATE_NEW_ALL_VISIBLE_CLEARED
+                            | pg_constants::XLH_UPDATE_OLD_ALL_VISIBLE_CLEARED)
+                        && decoded.blocks.len() > 1
+                        && pg_constants::HEAPBLK_TO_MAPBLOCK(decoded.blocks[0].blkno)
+                            == pg_constants::HEAPBLK_TO_MAPBLOCK(decoded.blocks[1].blkno)
+                    {
+                        timeline.put_wal_record(
+                            lsn,
+                            RelishTag::Relation(RelTag {
+                                forknum: pg_constants::VISIBILITYMAP_FORKNUM,
+                                spcnode: decoded.blocks[0].rnode_spcnode,
+                                dbnode: decoded.blocks[0].rnode_dbnode,
+                                relnode: decoded.blocks[0].rnode_relnode,
+                            }),
+                            pg_constants::HEAPBLK_TO_MAPBLOCK(decoded.blocks[0].blkno),
+                            ZenithWalRecord::ClearVisibilityMapFlags {
+                                new_heap_blkno: decoded.blocks[0].blkno,
+                                old_heap_blkno: decoded.blocks[1].blkno,
+                                flags: pg_constants::VISIBILITYMAP_VALID_BITS,
+                            },
+                        )?;
+                    } else {
+                        if (xlrec.flags & pg_constants::XLH_UPDATE_NEW_ALL_VISIBLE_CLEARED) != 0 {
+                            timeline.put_wal_record(
+                                lsn,
+                                RelishTag::Relation(RelTag {
+                                    forknum: pg_constants::VISIBILITYMAP_FORKNUM,
+                                    spcnode: decoded.blocks[0].rnode_spcnode,
+                                    dbnode: decoded.blocks[0].rnode_dbnode,
+                                    relnode: decoded.blocks[0].rnode_relnode,
+                                }),
+                                pg_constants::HEAPBLK_TO_MAPBLOCK(decoded.blocks[0].blkno),
+                                ZenithWalRecord::ClearVisibilityMapFlags {
+                                    old_heap_blkno: decoded.blocks[0].blkno,
```

Comment:
I would prefer to set old_heap_blkno to InvalidBlockNumber, if it's not used. That would be more clear, if you debug the log later (same with new_heap_blkno, of course)

Reply:
Sorry, but I prefer not to do it.
First of all because there is no InvalidBlockNumber: all blocks from 0 till u32::MAX are valid. Certainly we can "reserve" one block for our own needs, but ... if it is possible not to do it, it is better not reduce max table size  even on 8kb.
Certainly we can use Option, but IMHO it is overkill.
Frankly speaking I do not thin that this extra checks will make code more easy to understand and less error prone.
And from performance point of view (completely not important in this case), redundant assignment of bits is better than performing extra check.

Reply:
> Sorry, but I prefer not to do it.
> First of all because there is no InvalidBlockNumber: all blocks from 0 till u32::MAX are valid. Certainly we can "reserve" one block for our own needs, but ... if it is possible not to do it, it is better not reduce max table size even on 8kb.

https://github.com/postgres/postgres/blob/82331ed4dd60ce1f8db67e189f793f27cd1aac96/src/include/storage/block.h#L33

Reply:
O, really:) 
Now I remember that I have already see this constant in Postgres sources.

But still I think that if you prefer readability to compactness+efficiency (both are absolutely not critical in this case), then using Option is better than adding INVALID_BLOCK_ID and comparing it.